modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/CefParser.java

```diff
@@ -0,0 +1,449 @@
+                    case 6:
+                        event.setSeverity(headerFields.get(6));
+                        break;
+                }
+            }
+            String extensionString = cefString.substring(extensionStart);
+            Map<String, String> extensions = parseExtensions(extensionString);
+
+            if (removeEmptyValue) {
+                removeEmptyValue(extensions);
+            }
+            event.setExtensions(extensions);
+
+            // Translate possible ECS fields and remove them from extensions
+            Map<String, Object> translatedFields = extensions.entrySet()
+                .stream()
+                .filter(entry -> FIELD_MAPPING.containsKey(entry.getKey()))
+                .collect(Collectors.toMap(entry -> FIELD_MAPPING.get(entry.getKey()), entry -> {
+                    Class<?> fieldType = ECSFieldWithType.getFieldType(FIELD_MAPPING.get(entry.getKey()));
+                    return convertValueToType(entry.getValue(), fieldType);
+                }));
+
+            // Remove the translated entries from extensions
+            event.removeMappedExtensions();
+
+            ingestDocument.setFieldValue(targetField, event.toObject());
+            // Add ECS translations to the root of the document
+            if (translatedFields.isEmpty() == false) {
+                translatedFields.forEach(ingestDocument::setFieldValue);
+            }
+        } else {
+            throw new IllegalArgumentException("Invalid CEF format");
+        }
+    }
+
+    private Object convertValueToType(String value, Class<?> type) {
+        if (type == Long.class) {
+            return Long.parseLong(value);
+        } else if (type == Double.class) {
+            return Double.parseDouble(value);
+        } else if (type == Integer.class) {
+            return Integer.parseInt(value);
+        } else {
+            return value; // Default to String
+        }
+    }
+
+    private Map<String, String> parseExtensions(String extensionString) {
+        Map<String, String> extensions = new HashMap<>();
+        Matcher matcher = EXTENSION_NEXT_KEY_VALUE_PATTERN.matcher(extensionString);
+        int lastEnd = 0;
+        while (matcher.find()) {
+            String key = matcher.group(1);
+            String value = matcher.group(2);
+
+            // Convert extension field name to strict legal field_reference
+            if (key.endsWith("]")) {
+                key = convertArrayLikeKey(key);
+            }
+
+            extensions.put(key, desanitizeExtensionVal(value.trim()));
+            lastEnd = matcher.end();
+        }
+        // If there's any remaining unparsed content, throw an exception
+        if (lastEnd < extensionString.length()) {
+            throw new IllegalArgumentException("Invalid extensions; keyless value present: " + extensionString.substring(lastEnd));
+        }
+        return extensions;
+    }
+
+    private void removeEmptyValue(Map<String, String> map) {
```

Comment:
Methods that do not need access to instance members should be marked as `static`.